### ISSUE TITLE
should shaders be checked for execution-uniformity (e.g. barrier only in uniform control flow)

### ISSUE THREAD

**Issue by dneto0** [gpuweb/gpuweb#264]
Should WebGPU require an error when it can analyze erroneous non-uniformity?  

For example, is the compiler is obliged to throw an error in provably non-uniform cases, where a primitive is misused such as barrier or derivative.
Should the compiler be conservative and only permit provably good cases?

There are many options in the design space, especially since the dynamic behaviour of the shader invocations cannot in general be determined statically. (e.g. the halting problem).

This was discussed but unresolved in the San Diego F2F, notes are at this bookmark: https://docs.google.com/document/d/1tXp8CJ0GQa4TGS_4skRPegzFut9alZkRt4KCP7m7oQk/edit#bookmark=id.940m37e17g4l 

cc: @RobinMorisset

**Comment by dneto0** [comment]
@robinmorisset prototyped a conservative analysis of uniformity in WHLSL.
It was removed in part due to excessive complexity (?)

The delta is captured here: https://github.com/dneto0/WHLSL/compare/with-uniformity-analysis...dneto0:without-uniformity-analysis?expand=1 where the _old_ version has the uniformity analysis.

I've reviewed major parts of it and I tend to agree that the static analysis is probably not worth the complexity.  I also find it hard to follow.  If I read it correctly, it doesn't yet have a rule for how uniformity is treated at a function call boundary.

I think this is still quite open for discussion, however.

**Comment by RobinMorisset** [comment]
The rule I had for function calls at the time was basically to inline them. A smarter/more modular/more performant approach is likely possible, by essentially inverting the question: instead of asking "is this function valid in uniform/non-uniform control-flow, and with its arguments being this combination of uniform/non-uniform", we could ask "what is the minimum level of uniformity for its arguments/outside control-flow such that it is valid".
But this would require a significant refactoring of the spec/analysis, and would probably make it even more complex. So I instead removed the entire analysis.
I could try writing an updated version of the analysis with this approach if you think it would be useful. But if the current analysis is already considered too complex, I don't want to waste time doing this.

**Comment by grorg** [comment]
Discussed at the [8 April 2019 meeting](https://docs.google.com/document/d/1UiIZ7G7iruGYntiLSmhc-wRPj1sN1CqNcPdgeG51r1U/edit)

**Comment by dneto0** [comment]
Discussed in [2019-05-16 meeting](https://docs.google.com/document/d/1HR9vPk48-lhzHAsyF57QzjKMmi0gQHfSXMmyqZuFWkc/edit#):

- Agreed to require an error, consistently across implementations
- The analysis needs to be conservative: When it says "program is ok" it must be ok for all implementations.  It must start with a simple algorithm, and ok to reject tricky but ok cases.
